### PR TITLE
#41 Added initial version of the LatexFormula

### PR DESCRIPTION
**Pull request by johan-tuls**
## Description

In order to represent the formula in whatever context as desired. 
The formula should have a standard way of representing itself. Most likely in Latex format.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

## Checklist:

- [x] I have added tests that prove my fix is effective or that my feature works
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by egarciamendez — COMMENTED** [review]
I like this aproach but I am very curious to see what @bro-wi is cooking.

I would only like to add a reference to the formula label at the end. And maybe treat this object as a str in the same way we did with float and Formula. But that is more of a suggestion.

**Comment by johan-tuls** [comment]
> I was planning to spend some time next week on this subject, (although my mind is constantly prototyping). So I hope you're not in too much of a hurry ;-)

Not in a hurry ^_^ But i think once this feature has been implemented, the formula class can have really added value and it would be great to start using these formulas outside of blueprints.

**Comment by bro-wi** [comment]
Have been trying to cook something, but it quickly becomes very (and maybe unnecessarily) complex. For a straightforward one-line formula we can make it work by overiding the __add__,__sub__, __mul__,__div__, __pow__ . you have to write some logic to handle adding parenthesis (which quickly becomes quite complex), but it could work. a small (and way to simple) example:

```
@dataclass(frozen=True)
class LatexFormula:
    return_symbol: str
    result: float
    equation: str = ""
    numeric_equation: str = ""

    @property
    def complete(self) -> str:
        all_sub_equations = [self.return_symbol, self.equation, self.numeric_equation, str(self.result)]
        return " = ".join([eq for eq in all_sub_equations if eq != ""])

    @property
    def short(self) -> str:
        return f"{self.return_symbol} = {self.result}"

    def __type_check(self, other):
        if not isinstance(other, LatexFormula):
            raise TypeError(f"can't add LatexFormula to non-LatexFormula of type '{type(other)}'")

    def __str__(self):
        return self.complete

    def __add__(self, other):
        self.__type_check(other)

        return LatexFormula(
            "", self.result + other.result, f"{self.equation} + {other.equation}", f"{self.numeric_equation}+{other.numeric_equation}"
        )

    def __mul__(self, other):
        self.__type_check(other)

        # TODO: write check so parentheses are added when self or other is a summation, but not when there are already parentheses around them:
        # e.g.: a+b * c+d should be (a+b)*(c*d)
        #       (a+b)*c is already okay
        #       a+b * c should be (a+b)*c

        return LatexFormula(
            "", self.result * other.result, f"({self.equation})({other.equation})", f"({self.numeric_equation})({other.numeric_equation})"
        )
```

would work like this:

image


--> but then: 
-what will happen for multiline formulas? 
-And how would we include the formula label?
-Another thing I cant really tackle (yet) is inequality (the current implementation also fails on that).

so for the time being: I think have to KISS. I'm afraid that opting something very complex now will turn out to be more a hassle than we just 'type in' the result we want. It takes more effort of the developers, but I think it is more foolproof and flexible. Sorry for taking so long to come to this conclusion.. there were some escalations that required my attention.

**Comment by egarciamendez** [comment]
Task remaining before merging:

- [ ] Add docstrings
- [ ] Add tests

**Comment by bro-wi** [comment]
> Tasks remaining before merging:
> 
> * [ ]  Add docstrings
> * [ ]  Add tests

*[ ] fix for inequality
*[ ] add formula label

File: blueprints/codes/latex_formula.py
Comment (bro-wi) [inline blueprints/codes/latex_formula.py:14]: this will (probably) not work for inequalities?
+ include formula label

File: blueprints/codes/latex_formula.py
Comment (bro-wi) [inline blueprints/codes/latex_formula.py:9]: add docstrings  (also mention that the units are included in the numeric equation )

**Comment by johan-tuls** [comment]
> Glad we have almost reached a conclusion. @johan-tuls Can you add this to your solution?
> 
> ```python
>     def __str__(self):
>         return self.complete
> ```

Done

File: blueprints/codes/latex_formula.py
Comment (johan-tuls) [reply blueprints/codes/latex_formula.py:14]: I have added the formula label as comparison_operator_label with = as default.

**Comment by johan-tuls** [comment]
I have added the open requests. Let's see if this works out for the initial version.
@bro-wi @egarciamendez could you review the changes?

File: blueprints/codes/latex_formula.py
Comment (egarciamendez) [inline blueprints/codes/latex_formula.py:26]: I think this should be a float instead. Transforming to a str is the responsibility of the class itself.

File: blueprints/codes/latex_formula.py
Comment (egarciamendez) [inline blueprints/codes/latex_formula.py:1]: What do we do with inequalities? This class doesn't cover those situations.

File: tests/codes/test_latex_formula.py
Comment (egarciamendez) [inline tests/codes/test_latex_formula.py:9]: Use @pytest.fixtures instead and put the fixture outside of the testing class.

File: tests/codes/test_latex_formula.py
Comment (egarciamendez) [inline tests/codes/test_latex_formula.py:29]: The expected result has an extra unnecessary E -> "E = mc^2 =...

**Review by egarciamendez — CHANGES_REQUESTED** [review]
We're almost there 💪

File: tests/codes/test_latex_formula.py
Comment (johan-tuls) [reply tests/codes/test_latex_formula.py:29]: Fixed, the initial fixture was wrong

File: tests/codes/test_latex_formula.py
Comment (johan-tuls) [reply tests/codes/test_latex_formula.py:9]: done, great improvement!

File: docs/source/conf.py
Comment (johan-tuls) [reply docs/source/conf.py:7]: reversed

File: blueprints/codes/latex_formula.py
Comment (johan-tuls) [reply blueprints/codes/latex_formula.py:26]: as it only contains strings and in theorie should be possible for any formula (including string results) i suggest we stick to str for all subresults.

File: blueprints/codes/latex_formula.py
Comment (johan-tuls) [reply blueprints/codes/latex_formula.py:1]: most of them can be rewritten as formulas. Otherwise we might need a new object

**Review by egarciamendez — APPROVED** [review]
I think this should be the way to go. 🚀

**Comment by egarciamendez** [comment]
Let's wait for @bro-wi 's review to merge.

File: blueprints/codes/latex_formula.py
Comment (johan-tuls) [reply blueprints/codes/latex_formula.py:42]: Thanks for the review. 
I indeed think inequalities might need a different approach. The problem is that most inequalities aren't really inequalities, but just formula written in a different format so it is more readable for the end-user. 
So we might have to debate what would be the best approach.

However i think that for generic formula (y = .....x ...) we should try to complete the PR so we can start implementing Blueprints in regular apps and giving some momentum to the devlopment.

Comment (bro-wi) [reply blueprints/codes/latex_formula.py:42]: agree! We already have issue #50 to keep the inequalities on our radar, so let's merge and move on 🚀

**Pull request merged**
